<p>Q1. Show that a simple perceptron cannot solve the XOR problem from the given figure</p>
<p>Equation of a perceptron with two inputs is given by:</p> 
<p><center><b>w0 +w1X1 +w2X2 =0</b></center></p>


In [9]:
import numpy as np

# BINARY STEP FUNCTION - threshold = 0
def binary(x):
    return 1 if x >= 0 else 0

# BIPOLAR STEP FUNCTION - threshold = 0
def bipolar(x):
    return 1 if x >= 0 else -1

# BINARY SIGMOIDAL FUNCTION
def bin_sigmoid(x):
    return 1/(1 + np.exp(-x))

# BIPOLAR SIGMOIDAL FUNCTION
def bip_sigmoid(x):
    return 2/(1 + np.exp(-x)) - 1

In [10]:
# import the module numpy
import numpy as np

inputs = np.array(
    [
        [-1, 1],
        [1, 1],
        [1, -1],
        [-1, -1]
    ])
 
output = np.array([1, -1, 1, -1])

# initialise weights, bias , learning rate, epoch
weight = [0, 0]
bias = 0
learning_rate = 1
epoch = 100

In [11]:
for i in range(epoch):
    print("\n\n\n Epoch : ", i+1)
    for j in range(inputs.shape[0]):
        target = output[j]
        x1 = inputs[j][0]
        x2 = inputs[j][1]
        print("Inputs: ", x1, x2)
        print("Target: ", target)
        y = (x1 * weight[0]) + (x2 * weight[1]) + bias

        y_out = bipolar(y)
        print("y = ",y)
        print("y_out = ", y_out)
        print("W1: ", weight[0])
        print("W2: ", weight[1])
        print("Bias: ", bias)
        
        # updation of weights and bias
        weight[0] += learning_rate * target * x1
        weight[1] += learning_rate * target * x2
        bias += learning_rate * target




 Epoch :  1
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  2
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  3
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  4
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1


y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  59
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  60
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_out =  -1
W1:  -2
W2:  0
Bias:  0
Inputs:  -1 -1
Target:  -1
y =  3
y_out =  1
W1:  -1
W2:  -1
Bias:  1



 Epoch :  61
Inputs:  -1 1
Target:  1
y =  0
y_out =  1
W1:  0
W2:  0
Bias:  0
Inputs:  1 1
Target:  -1
y =  1
y_out =  1
W1:  -1
W2:  1
Bias:  1
Inputs:  1 -1
Target:  1
y =  -2
y_

<h2>Inference</h2>

As we can see that , single percpetron can not solve the XOR in 100 epochs. The weights after 100 epoch is w1:-1 w2:-1 B:1 and the y is 3

<h3> Q2. Show that a two-layer perceptron can solve the XOR problem </h3>

In [12]:
#Initialising the Inputs 
inputs = np.array([[-1,1],
                  [1,1],
                  [-1,-1],
                  [1,-1]])
target = np.array([1,-1,1,-1])
weight = np.array([[0.1,0.1],
                   [0.1,0.1]])
bias_1 = 0.2
bias_2 = 0.1
v1 = v2 = bias_3 =0.5
alpha = 0.5

In [14]:
#Defining Bipolar Activation function
def bipolar(z_in):
    if z_in<0:
      return -1
    else:
      return 1


num_epoch = 1
while(num_epoch<4):   
    print("\n\n Epoch : ", num_epoch,"\n-----------")
    for i in range(len(inputs)):    
        w_sum = np.dot(inputs[i],weight)
        zin_1 = w_sum[0] + bias_1  
        z1 = bipolar(zin_1)   
        zin_2 = w_sum[1] + bias_2
        z2 = bipolar(zin_2)   
        
        yin = z1*v1 + z2*v2 + bias_3
        y = bipolar(yin)  

        if target[i]!=y:    
            close_zero = np.abs(np.array([zin_1,zin_2])).argmin()  
            if(target[i] == 1 and close_zero == 0)or(target[i] == -1 and zin_1>=0):   
                weight[0][0] = weight[0][0] + alpha*(target[i]-zin_1)*inputs[i][0]
                weight[1][0] = weight[1][0] + alpha*(target[i]-zin_1)*inputs[i][1]
                bias_1 = bias_1 + alpha*(target[i]-zin_1)
            if(target[i] == 1 and close_zero == 1)or(target[i] == -1 and zin_2>=0):   
                weight[0][1] = weight[0][1] + alpha*(target[i]-zin_2)*inputs[i][0]
                weight[1][1] = weight[1][1] + alpha*(target[i]-zin_2)*inputs[i][1]
                bias_2 = bias_2 + alpha*(target[i]-zin_2)
    print("Weight:\n",weight,"\nBias:\n",np.array([bias_1, bias_2]) )
    num_epoch = num_epoch + 1

print("\n\nFinal Weights:\n--------------\n",weight,"\nBias:\n---------\n",np.array([bias_1,bias_2]))



 Epoch :  1 
-----------
Weight:
 [[-1.35 -0.55]
 [ 0.15 -0.55]] 
Bias:
 [ 0.25 -0.55]


 Epoch :  2 
-----------
Weight:
 [[-1.35 -0.55]
 [ 0.15 -0.55]] 
Bias:
 [ 0.25 -0.55]


 Epoch :  3 
-----------
Weight:
 [[-1.35 -0.55]
 [ 0.15 -0.55]] 
Bias:
 [ 0.25 -0.55]


Final Weights:
--------------
 [[-1.35 -0.55]
 [ 0.15 -0.55]] 
Bias:
---------
 [ 0.25 -0.55]


<h2>Testing model</h2> 

In [15]:
print("XOR GATE ")
x1 = int(input("Enter the Input 1: "))
x2 = int(input("Enter the Input 2: "))
inputs = np.array([x1,x2])

w_sum = np.dot(inputs,weight)
zin_1 = w_sum[0] + bias_1
z1 = bipolar(zin_1)
zin_2 = w_sum[1] + bias_2
z2 = bipolar(zin_2)
yin = z1*v1 + z2*v2 + bias_3
y = bipolar(yin)
print("The output : ",y)

XOR GATE 
Enter the Input 1: 1
Enter the Input 2: 1
The output :  -1


<h2>Inference</h2>

Madaline Network solves the XOR in 3 epochs

 Final Weights and bias: W11 = -1.35, W12 = -0.55, W21 = 0.25, W22 = -0.55, Bias 1 = 0.25, Bias 2 = -0.55